### ISSUE TITLE
[3.5.0] app.toTree not outputting broccoli-funnel anymore?

### ISSUE THREAD

**Issue by ffaubry** [ember-cli/ember-cli#8128]
Here are the steps:
New project: `ember new xxx`
In project:
`npm i broccoli-funnel --save-dev`
`npm i moment-timezone --save`
Edit ember-cli-build.js:
`const Funnel = require('broccoli-funnel');`
...
```
  const momentTimezone = new Funnel('node_modules/moment-timezone/builds', {
    srcDir: '/',
    include: ['moment-timezone-with-data.min.js'],
    destDir: 'assets'
  });

  return app.toTree([momentTimezone]);
```
Build:
`ember b -dev`
Check dist folder:
![image](https://user-images.githubusercontent.com/3506166/46967096-68e5d580-d064-11e8-9821-d642a4d5d4d2.png)

No `moment-timezone-with-data.min.js`

I tried with other resources (jpegs, css, etc.), same deal.

---

Output from `ember version --verbose && npm --version && yarn --version`:
```
ember-cli: 3.5.0
http_parser: 2.8.0
node: 10.11.0
v8: 6.8.275.32-node.28
uv: 1.23.0
zlib: 1.2.11
ares: 1.14.0
modules: 64
nghttp2: 1.33.0
napi: 3
openssl: 1.1.0i
icu: 62.1
unicode: 11.0
cldr: 33.1
tz: 2018e
os: darwin x64
6.4.1
1.10.1
```

**Comment by Turbo87** [comment]
@ffaubry with 3.4.x this worked and with 3.5.0 this fails, correct?

**Comment by rwjblue** [comment]
@ffaubry - Can you share a specific reproduction that we can clone for testing/verification?

**Comment by ffaubry** [comment]
@rwjblue 
`ember s` works. The issue is in building. I'm trying to funnel a file in the dist folder.

You said: "Can you share a specific reproduction that we can clone for testing/verification"
I made the smaller set of reproducible steps.
To summarize, my steps works with 3.4.3 and fail with 3.5. Same machine. Just different ember-cli.
I'm not sure what else you need to repro this issue?
 
As for your request, I ran:
```
EMBER_CLI_BROCCOLI_2=false ember b -dev
```

Same issue, my file has not been funneled:
![image](https://user-images.githubusercontent.com/3506166/46974943-38a93180-d07a-11e8-9ca1-390a06ac3686.png)

**Comment by Turbo87** [comment]
@ffaubry can you check if any of the 3.5.0 beta releases also show the similar issues?

**Comment by ffaubry** [comment]
@Turbo87 
```
ember-cli: 3.5.0-beta.2
```

Same steps. Same issue.

**Comment by Turbo87** [comment]
okay, does `beta.1` fix it or is that also broken?

**Comment by ffaubry** [comment]
@Turbo87 (I'm not doing a good job at following your request...)
```
ember-cli: 3.5.0-beta.1
```
Same steps. Same issue.

Summary: It seems to be broken on 3.5.0, 3.5.0-beta.1, and 3.5.0-beta.2

Anything else?

**Comment by Turbo87** [comment]
> Anything else?

That already narrows it down quite a bit, thanks a lot! :)

**Comment by wagenet** [comment]
I may be seeing the same issue. After updating ember-cli to 3.5, my funnel'ed stuff stopped working as well. Here's my yarn.lock change as well as my ember-cli-build.js: https://gist.github.com/wagenet/52a7b8f09c13fbf0454af2b4ef986857#file-ember-cli-build-js-L116

**Comment by rwjblue** [comment]
Working on this with @wagenet, I think we have a path forward...

**Comment by wagenet** [comment]
At @rwjblue's direction I modified `node_modules/ember-cli/lib/broccoli/ember-app.js`'s `toTree` method to the following:

```js
  toTree(additionalTrees) {
    let packagedTree;
    let trees = this.toArray();
    let packageFn = this.options.package;

    let fullTree = mergeTrees(trees, {
      overwrite: true,
      annotation: 'Full Application',
    });

    fullTree = this._debugTree(fullTree, 'prepackage');

    if (isExperimentEnabled('PACKAGER')) {
      if (this._isPackageHookSupplied) {
        packagedTree = packageFn.call(this, fullTree);
      } else {
        this.project.ui.writeWarnLine('`package` hook must be a function, falling back to default packaging.');
      }
    } else {
      packagedTree = this._legacyPackage(fullTree);
    }
    let MergeTreesConst = require('broccoli-merge-trees');
    let combinedPackageTree = new MergeTreesConst([].concat(packagedTree, additionalTrees).filter(Boolean));

    return this.addonPostprocessTree('all', combinedPackageTree);
  }
```

This appeared to resolve my issue. I also realized that my Funnel was outputting to `/images` instead of `/public/images`. When I updated to use `/public/images` things worked correctly without the above changes. My understanding from @rwjblue is that both should work, so we still should make this (or a similar) correction.

**Comment by linearza** [comment]
Updating my output dir to `'public/assets/images'`seems fix it for me, thanks! @wagenet

**Comment by rwjblue** [comment]
RE: @linearza's https://github.com/ember-cli/ember-cli/issues/8128#issuecomment-430258434

<details><summary>That warning in the console is unrelated to this bug.</summary> The issue there is the migration to use Broccoli v2.0.0. That migration moves the various `tmp` dirs created by the build out of the project's own `tmp/` directory and into your default system temp dir, and it also drops support for older/legacy style plugins that don't extend from `broccoli-plugin` base class.

You can confirm this by manually disabling the Broccoli 2 experiment and rebuilding:

```
EMBER_CLI_BROCCOLI_2=false ember s
```

In your specific case (`TemplateLinter` plugin), I think you need to update ember-cli-deprecation-workflow to 1.0.0.
</details>
<br>
The bug being worked on here is specifically about the arguments to `toTree` being ignored unless the contents of those trees are in "special" directories.

**Comment by ffaubry** [comment]
@rwjblue @wagenet @Turbo87 
Following the instructions above, I confirm that updating my `distDir` in the Funnel options to include the path `\public` (as `/public/xxx`) fixed my issue.

Thanks!

**Comment by makepanic** [comment]
I've seen the same issue.
The project used a Funnel like this:
```
new Funnel('public/images/icons', {
        srcDir: 'default',
        destDir: 'images/icons'
      })
```
changing it to
```
new Funnel('public/images/icons/', {
        srcDir: 'default',
        destDir: 'public/images/icons'
      })
```
solved the issue. Behavior looks also to be the same as it was before 3.5

**Comment by john-griffin** [comment]
Just hit this issue on a `3.5` app too. We copy a `_redirects` file for netlify configuration and got it working by changing:

``` js
new Funnel('./', {
  files: ['_redirects']
});
```

to:

``` js
new Funnel('./', {
  files: ['_redirects'],
  destDir: 'public'
});
```

**Comment by tomwayson** [comment]
As an addon author, I can't just prefix the `destDir` w/ `'public'` in `treeForPublic()` b/c I don't know what version of ember the app is using.

Is there any guidance from the core team as to how addon authors should handle this?

**Comment by wagenet** [comment]
@tomwayson my understanding is that its a bug and should be fixed. Not sure if anyone is actually working on fixing it yet. @rwjblue

**Comment by wagenet** [comment]
Fix in #8150
